### PR TITLE
scaleway: allocatable ip

### LINKED ISSUE #3387: scaleway
### Summary

Dear,

I tried to use `scaleway-cli` through ansible without succss
Indeed with following ansible statements
```yaml
      scaleway_compute:
        name: 'someone.com'
        state: 'present'
        # Roxky Linux 8
        image: '3fdb4082-cab9-4802-ae0b-3c18dcc59281'
        organization: 'someone'
        region: 'par1'
        commercial_type: 'DEV1-M'
        public_ip: 'XX.YY.ZZ.AA'
        security_group: 'xxxx-xxx-xxxx-xxxxx'
        wait: true
```
An error message is raised

``` 

The full traceback is:
Traceback (most recent call last):
  File "/home/jonathan/.ansible/tmp/ansible-tmp-1631547906.8553545-19198-189669113758992/AnsiballZ_scaleway_compute.py", line 100, in <module>
    _ansiballz_main()
  File "/home/jonathan/.ansible/tmp/ansible-tmp-1631547906.8553545-19198-189669113758992/AnsiballZ_scaleway_compute.py", line 92, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/home/jonathan/.ansible/tmp/ansible-tmp-1631547906.8553545-19198-189669113758992/AnsiballZ_scaleway_compute.py", line 40, in invoke_module
    runpy.run_module(mod_name='ansible_collections.community.general.plugins.modules.scaleway_compute', init_globals=dict(_module_fqn='ansible_collections.community.general.plugins.modules.scaleway_compute', _modlib_path=modlib_path),
  File "/usr/lib/python3.9/runpy.py", line 210, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File "/usr/lib/python3.9/runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/usr/lib/python3.9/runpy.py", line 87, in _run_code
    exec&#40;code, run_globals&#41;
  File "/tmp/ansible_scaleway_compute_payload_qqsr2ljv/ansible_scaleway_compute_payload.zip/ansible_collections/community/general/plugins/modules/scaleway_compute.py", line 673, in <module>
  File "/tmp/ansible_scaleway_compute_payload_qqsr2ljv/ansible_scaleway_compute_payload.zip/ansible_collections/community/general/plugins/modules/scaleway_compute.py", line 669, in main
  File "/tmp/ansible_scaleway_compute_payload_qqsr2ljv/ansible_scaleway_compute_payload.zip/ansible_collections/community/general/plugins/modules/scaleway_compute.py", line 641, in core
TypeError: 'NoneType' object is not iterable
fatal: [localhost]: FAILED! => {
    "changed": false,
    "module_stderr": "Traceback (most recent call last):\n  File \"/home/jonathan/.ansible/tmp/ansible-tmp-1631547906.8553545-19198-189669113758992/AnsiballZ_scaleway_compute.py\", line 100, in <module>\n    _ansiballz_main()\n  File \"/home/jonathan/.ansible/tmp/ansible-tmp-1631547906.8553545-19198-189669113758992/AnsiballZ_scaleway_compute.py\", line 92, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/home/jonathan/.ansible/tmp/ansible-tmp-1631547906.8553545-19198-189669113758992/AnsiballZ_scaleway_compute.py\", line 40, in invoke_module\n    runpy.run_module(mod_name='ansible_collections.community.general.plugins.modules.scaleway_compute', init_globals=dict(_module_fqn='ansible_collections.community.general.plugins.modules.scaleway_compute', _modlib_path=modlib_path),\n  File \"/usr/lib/python3.9/runpy.py\", line 210, in run_module\n    return _run_module_code(code, init_globals, run_name, mod_spec)\n  File \"/usr/lib/python3.9/runpy.py\", line 97, in _run_module_code\n    _run_code(code, mod_globals, init_globals,\n  File \"/usr/lib/python3.9/runpy.py\", line 87, in _run_code\n    exec&#40;code, run_globals&#41;\n  File \"/tmp/ansible_scaleway_compute_payload_qqsr2ljv/ansible_scaleway_compute_payload.zip/ansible_collections/community/general/plugins/modules/scaleway_compute.py\", line 673, in <module>\n  File \"/tmp/ansible_scaleway_compute_payload_qqsr2ljv/ansible_scaleway_compute_payload.zip/ansible_collections/community/general/plugins/modules/scaleway_compute.py\", line 669, in main\n  File \"/tmp/ansible_scaleway_compute_payload_qqsr2ljv/ansible_scaleway_compute_payload.zip/ansible_collections/community/general/plugins/modules/scaleway_compute.py\", line 641, in core\nTypeError: 'NoneType' object is not iterable\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error",
    "rc": 1
```

Moreover in order to use this ansible module we have define the `SCW_TOKEN` variable nonetheless this variable is deprecated and scaleway use  now ` SCW_SECRET_KEY`

Thanks to keep updated this module

best regards

### Issue Type

Bug Report

### Component Name

scaleway_compute

### Ansible Version

```console (paste below)
$ ansible --version
ansible [core 2.11.4] 
  config file = None
  configured module search path = ['/home/jonathan/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/jonathan/.venv/ansible/lib/python3.9/site-packages/ansible
  ansible collection location = /home/jonathan/.ansible/collections:/usr/share/ansible/collections
  executable location = /home/jonathan/.venv/ansible/bin/ansible
  python version = 3.9.5 (default, May 11 2021, 08:20:37) [GCC 10.3.0]
  jinja version = 3.0.1
  libyaml = True


```

### Configuration

```console (paste below)
$ ansible-config dump --only-changed

```

### OS / Environment

_No response_

### Steps to Reproduce

```yaml
      scaleway_compute:
        name: 'someone.com'
        state: 'present'
        # Roxky Linux 8
        image: '3fdb4082-cab9-4802-ae0b-3c18dcc59281'
        organization: 'someone'
        region: 'par1'
        commercial_type: 'DEV1-M'
        public_ip: 'XX.YY.ZZ.AA'
        security_group: 'xxxx-xxx-xxxx-xxxxx'
        wait: true
```

### Expected Results

I expect to not get this error

### Actual Results

```console (paste below)

```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by bioinfornatics**
##### SUMMARY

Fixes  #3387 

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

scaleway

### CODE REVIEW

**Comment by russoz** [comment]
Hi @bioinfornatics glad that you took the leap :-) Thank you for the contribution! Please add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to).

You can run the sanity tests locally, before committing. Give it a read on https://github.com/ansible-collections/community.general/blob/main/CONTRIBUTING.md for details.

A shortcut you may want to try is to use this tool written by yours truly, `andebox`. If you `pip install andebox ansible-core` (or `ansible`, depending on what you are testing), you can then simply run it in the c.g. repo root:
`$ andebox test -- sanity --docker default --python 3.8 plugins/modules/cloud/scaleway/scaleway_compute.py`
(the docker and python settings are optional).

File: plugins/module_utils/scaleway.py [hunk 1]
```diff
@@ -13,7 +13,7 @@
 
 def scaleway_argument_spec():
     return dict(
-        api_token=dict(required=True, fallback=(env_fallback, ['SCW_TOKEN', 'SCW_API_KEY', 'SCW_OAUTH_TOKEN', 'SCW_API_TOKEN']),
+        api_token=dict(required=True, fallback=(env_fallback, ['SCW_TOKEN', 'SCW_API_KEY', 'SCW_OAUTH_TOKEN', 'SCW_API_TOKEN', 'SCW_SECRET_KEY']),
```
Comment (felixfontein) [inline plugins/module_utils/scaleway.py:16]: This would rather be a feature than a bugfix. It would be better to add this in another PR, since with a new feature in it, it cannot be backported to the stable-2 branch.

**Comment by felixfontein** [comment]
@bioinfornatics would you mind adding a changelog fragment, and looking at the other comments? Thanks.

needs_info

**Comment by bioinfornatics** [comment]
Hi @all I very sorry to not be enough present.
I will try to apply all your suggestion soon

sorry again

Have a good day

**Comment by russoz** [comment]
@bioinfornatics I am not sure why you posted the sanity test output, the CI here give us that for every branch. The suggestion was for you to run the tests locally, before pushing, to help you identify problems earlier.

Thanks for the changelog fragment, but I believe there is still an unanswered comment from Felix. Could you please take a look? Thanks

**Comment by russoz** [comment]
> > $ andebox test -- sanity --docker default --python 3.8 plugins/modules/cloud/scaleway/scaleway_compute.py
> 
> FYI: never use `--python` with `sanity`. Otherwise you're skipping some tests and potentially run some others with the wrong Python interpreter.

Never is too strong a word. The advantage of running it with only one python version is that the speed - when changing things quickly and trying them, having shorter cycles is a good thing. But the risk you mentioned exists and we should run it without specifying python before committing.

Something similar could be said about Ansible itself: if we use ansible-test from core 2.11 only, we might have surprises when the CI tests that same code in 2.10 or 2.9 later on.

**Comment by bioinfornatics** [comment]
```
$ andebox test -- sanity  --python 3.9 plugins/modules/cloud/scaleway/scaleway_compute.py 
collection = community.general
directory  = /tmp/andebox.r8vgq5il/ansible_collections/community/general
Running sanity test 'action-plugin-docs' with Python 3.9
Running sanity test 'ansible-doc' with Python 3.9
Running sanity test 'changelog' with Python 3.9
Running sanity test 'compile' with Python 3.9
Running sanity test 'empty-init' with Python 3.9
Running sanity test 'future-import-boilerplate' with Python 3.9
Running sanity test 'ignores'
Running sanity test 'import' with Python 3.9
Running sanity test 'line-endings' with Python 3.9
Running sanity test 'metaclass-boilerplate' with Python 3.9
Running sanity test 'no-assert' with Python 3.9
Running sanity test 'no-basestring' with Python 3.9
Running sanity test 'no-dict-iteritems' with Python 3.9
Running sanity test 'no-dict-iterkeys' with Python 3.9
Running sanity test 'no-dict-itervalues' with Python 3.9
Running sanity test 'no-get-exception' with Python 3.9
Running sanity test 'no-illegal-filenames' with Python 3.9
Running sanity test 'no-main-display' with Python 3.9
Running sanity test 'no-smart-quotes' with Python 3.9
Running sanity test 'no-unicode-literals' with Python 3.9
Running sanity test 'pep8' with Python 3.9
Running sanity test 'pslint'
Running sanity test 'pylint' with Python 3.9
ERROR: Found 43 pylint issue(s) which need to be resolved:
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:191:36: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:194:14: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:199:29: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:200:36: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:206:14: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:210:33: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:213:41: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:231:37: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:250:14: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:257:36: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:282:14: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:288:41: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:308:37: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:311:14: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:321:39: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:323:14: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:350:39: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:371:35: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:379:22: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:388:18: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:391:31: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:413:39: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:419:33: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:423:39: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:427:18: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:457:26: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:465:33: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:470:39: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:477:18: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:504:26: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:510:35: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:518:18: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:526:18: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:549:14: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:568:29: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:569:29: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:573:29: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:591:20: use-dict-literal: Consider using {} instead of dict()
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:605:38: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:608:14: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:614:41: consider-using-f-string: Formatting a regular string which could be a f-string
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:651:13: use-dict-literal: Consider using {} instead of dict()
ERROR: plugins/modules/cloud/scaleway/scaleway_compute.py:662:23: use-dict-literal: Consider using {} instead of dict()
See documentation for help: https://docs.ansible.com/ansible/2.11/dev_guide/testing/sanity/pylint.html
Running sanity test 'replace-urlopen' with Python 3.9
Running sanity test 'runtime-metadata' with Python 3.9
Running sanity test 'shebang' with Python 3.9
Running sanity test 'shellcheck'
Running sanity test 'symlinks' with Python 3.9
Running sanity test 'use-argspec-type-path' with Python 3.9
Running sanity test 'use-compat-six' with Python 3.9
Running sanity test 'validate-modules' with Python 3.9
Running sanity test 'yamllint' with Python 3.9
ERROR: The 1 sanity test(s) listed below (out of 32) failed. See error output above for details.
pylint
``` 

Error below come from history python style coding and should not be  part of this feature request

**Comment by felixfontein** [comment]
I'm not sure why you are posting that log. The test runs in CI and passes there (everything's green).

**Comment by bioinfornatics** [comment]
> I'm not sure why you are posting that log. The test runs in CI and passes there (everything's green).

I do not know what is wrong.  I tried to help the scaleway's code to solve our enterprise use of this collection. So it would be better to close and cancel this merge request to let a maintainer fix it. Like this will be better for everybody.

As tell if this can not fixed we have to move to another cloud provider as nowaday deploy an infrastructure as code is the way to do.

So please do what you want, but fix the bug.

Thanks

**Comment by felixfontein** [comment]
Everything that's currently missing for this being able to get merged is summed up in https://github.com/ansible-collections/community.general/pull/3418#discussion_r725454002: your PR introduces a breaking change that must be removed (and replaced by a deprecation warning, in case this should eventually fail after a deprecation period). In that discussion you claimed to implemnent this, and then you claimed that it will issue a warning in that case, but you did not change the code, it still fails instead of emitting a warning.

**Comment by bioinfornatics** [comment]
> Everything that's currently missing for this being able to get merged is summed up in [#3418 (comment)](https://github.com/ansible-collections/community.general/pull/3418#discussion_r725454002): your PR introduces a breaking change that must be removed (and replaced by a deprecation warning, in case this should eventually fail after a deprecation period). In that discussion you claimed to implemnent this, and then you claimed that it will issue a warning in that case, but you did not change the code, it still fails instead of emitting a warning.

Indeed, when I reply, I did the change `fail_json` by `warn` and thought that was enough. I agree that in case someone try to use the same IP two create 2  instances with  `fail_json` or with `warn` that will fail.
I am not in favor to do some black magic as with follows scenario I show that is most of the time error-prone or instances have to be removed so the advantage given by the black magic is too low to my current understanding.

Below scenarios show 2 instances (a, b) with same IP with a black magic fix

Scenario 1:
> the IP attached to instance (a) is the one expected 
> the instance (b)  was not created as the black magic emit a warning without failure

Conclusion 1:
>  Created instance (a) own expected IP and is present  but the instance (b) is missing. End user could miss the warnings as the playbooks ends successfully with one of instance created as expect. Some time have  will be wasted until end-user see and fix the playbooks . The playbooks have to be fixed and the black magic did not help

Scenario 2 (worst case):

> the IP attached to instance (a) is not the one expected 
> the instance (b)  was not created as the black magic emit a warning without failure

Conclusion 2:

> Created instance (a) have to be removed,end  user could miss the warnings as the playbooks end successfully. It will discover that instance b is missing and take some time to understand the problem. So the playbooks have to be fixed and the black magic did not help

Scenario 2:
> the IP attached to instance (a) is the one expected 
> the instance (b)  was not created as the black magic emit a warning without failure

Conclusion 2:
>  Created instance (a) own expected IP and is present  but the instance (b) is missing. End user could miss the warnings as the playbooks ends successfully with one of instance created as expect. Some time have  will be wasted until end-user see and fix the playbooks . The playbooks have to be fixed and the black magic did not help

As I do not see in which case the black magic (to keep one  pair IP-instance and filter out others pairs)  is interesting, I am not able to patch it.

**Comment by felixfontein** [comment]
I'm well aware of the principle of "fail fast", and in fact I like it a lot. To me it looks like you are actively ignoring my point: **we try not to break backwards compatibility**. By failing in a situation that didn't fail before, you are doing exactly that. I did point out to you how we usually fix such situations: we add a deprecation warning, and switch from deprecation warning to failing when the deprecation expires. If that's not acceptable to you, then I'm really sorry, but we won't merge your patch.

**Comment by bioinfornatics** [comment]
I think the situation meet us to be confused. So if: 

1.   I let the `warn` statement 
2. add a deprecation warnings
3. add few line of the black magic
how did you spot thes black magic code ? is their a boolean or a constantant var to use ? 
such that
```
if ALLOW_DEPRECATED_LOGIC:
....
else
....
```

**Comment by bioinfornatics** [comment]
> > > $ andebox test -- sanity --docker default --python 3.8 plugins/modules/cloud/scaleway/scaleway_compute.py
> > 
> > 
> > FYI: never use `--python` with `sanity`. Otherwise you're skipping some tests and potentially run some others with the wrong Python interpreter.
> 
> Never is too strong a word. The advantage of running it with only one python version is that the speed - when changing things quickly and trying them, having shorter cycles is a good thing. But the risk you mentioned exists and we should run it without specifying python before committing.
> 
> Something similar could be said about Ansible itself: if we use ansible-test from core 2.11 only, we might have surprises when the CI tests that same code in 2.10 or 2.9 later on.

Is it ok for you @russoz ?

**Comment by russoz** [comment]
hi @bioinfornatics 

> I think the situation meet us to be confused. So if:
> 
> 1. I let the `warn` statement
> 2. add a deprecation warning
> 3. add few lines of the black magic
>    how did you spot this black magic code in order to ease the maintenance ?
>    is there a boolean or a constant var to use ?
>    such that

Sorry, you wrote many times about this alleged black magic but I don't really see anything odd or black magic-ish being requested or suggested.

The bottom line is, before we had this:
```python
    lookup = [ip["id"] for ip in ip_list]
    if public_ip in lookup:
        return {"public_ip": public_ip}
```
If `public_ip` appears once or more than once in the `lookup` list, the result will be the same.

With your change we have:
```python
    lookup = [ip for ip in ip_list if ip["id"] == public_ip]
    if len(lookup) == 1:
        if public_ip in lookup[0]["id"]:
            return {"public_ip": public_ip}
    else:
        msg = "Non unique IP address ID: %s, %d occurences found" % (public_ip, len(lookup))
        compute_api.module.fail_json(msg=msg)
```
Which returns the same result as before when `public_ip` appears only once in `lookup`, but it explodes in the face of the user if occurs twice or more (and to be fair, it blows up when it does not occur as well, which seems to be a new bug being introduced). What Felix has been trying to say is that, by default, the module should behave the same way as before (but generating a deprecation warning). You need a flag in the module argument_spec, say, `allow_multiple_ips` with `default=True`, then your code would look like:
```python
    lookup = [ip for ip in ip_list if ip["id"] == public_ip]
    if not lookup:
        return   # this is the old behaviour of the module, but probably a bug. Maybe return {} instead?
    if len(lookup) == 1:
        # if public_ip in lookup[0]["id"]: => this is redundant because all elements of lookup are already == public_ip
        return {"public_ip": public_ip}
    elif compute_api.module.params['allow_multiple_ips']:
        compute_api.module.deprecate("more than one IP found, this is deprecated and will generate an error in the future", version... )
        return {"public_ip": public_ip}
    else:  # allow_multiple_ips is False
        msg = "Non unique IP address ID: %s, %d occurences found" % (public_ip, len(lookup))
        compute_api.module.fail_json(msg=msg)
```

I hope it helps clarify things a little bit.

> And anyway I propose a solution to fix a bug so I would like only a fix with mine or another patch in order to use the scaleway cloud. I have no connection with this code, just fix it please. A fix helps more scaleway than me unitarly so no regret to have.
> 
> Lastly, I do not break backwards compatibility because until now this black magic code never existed. So even before this patch, the automation script lead to an error and exit when a same IP is given for two instances.

Your proposal is appreciated, we also have no direct connection to this specific piece of code. Fixes are always welcome, but our role as maintainers of the collection is to make sure that all contributions are safe and do not break existing code. As demonstrated above, that does not seem to be case, with the code in its current state.

You are claiming that previously the script would also lead to an error if having more than one IP in that list. Could you please show us where in the code that error would occur in that case? Maybe that can help us better understand what you are trying to say. As we are not familiar (I am not, and assuming Felix is on the same level) with Scaleway's API, nor with the details of the code, we are looking at this very localized change and making inferences out of it.

File: changelogs/fragments/3387-allocatable_ip.yml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - scaleway - allow to create instance with an allocatable IP (https://github.com/ansible-collections/community.general/pull/3418).
```
Comment (russoz) [inline changelogs/fragments/3387-allocatable_ip.yml:2]: maybe it is a (lack of) terminology thing on my side but, here the description mentions the *allocatable IP*, whilst in the code itself it seems to be related to *public IP*. Assuming that they both refer to the same thing, I would suggest a tiny wordsmithing here, maybe:
```suggestion
  - scaleway - allow to create instance with a public allocatable IP (https://github.com/ansible-collections/community.general/pull/3418).
```
Would that make sense (and perhaps be clearer to someone not as familiar with Scaleway)?

File: plugins/modules/cloud/scaleway/scaleway_compute.py [hunk 3]
```diff
@@ -256,9 +256,13 @@ def public_ip_payload(compute_api, public_ip):
...
         compute_api.module.fail_json(msg="Error in getting the IP information from: %s" % response.json)
 
-    lookup = [ip["id"] for ip in ip_list]
-    if public_ip in lookup:
-        return {"public_ip": public_ip}
+    lookup = [ip for ip in ip_list if ip["id"] == public_ip]
+    if len(lookup) == 1:
+        if public_ip in lookup[0]["id"]:
+            return {"public_ip": public_ip}
+    else:
```
Comment (russoz) [inline plugins/modules/cloud/scaleway/scaleway_compute.py:263]: The case where `len(lookup) == 0` seems like a bug to me.

**Comment by felixfontein** [comment]
I don't think it's necessary to add a new option for this, just deprecating the old behavior is fine (though the deprecation should not happen in a bugfix PR, this would be a violation of semver). So I would do:
1. Change this PR so that the old behavior is kept without failing or warning.
2. Create a follow-up PR which deprecates the current behavior if no or multiple IPs are found.

**Comment by russoz** [comment]
The idea of the flag is to give users the opportunity of having the new behaviour today - without getting a deprecation warning. It is a bit frustrating for the user to get a notification that you are using something that is deprecated and not begin able to do anything about it.

**Comment by bioinfornatics** [comment]
Thanks @felixfontein, @russoz, @remyleone, @sieben
As this bug and merge request of around 20 lines of codes is still open after 2 months I will close it.

Feel free to submit a fix.


Thanks for your works
I wish you a good day

**Comment by bioinfornatics** [comment]
Hi did someone has fixed it ?
